2.isyntax conversion error

raw2ometiff producing the following error when converting the raw output of isyntax2raw conversion of the test file 2.isyntax. Only seems to effect this specific file:

```
2020-03-27 10:29:44,651 [main] INFO  org.perf4j.TimingLogger - start[1585304984634] time[16] tag[getInputTileBytes]
2020-03-27 10:29:44,653 [main] INFO  org.perf4j.TimingLogger - start[1585304984651] time[1] tag[getInputTileBytes]
2020-03-27 10:29:44,664 [pool-1-thread-2] INFO  org.perf4j.TimingLogger - start[1585304984634] time[29] tag[writeTile]
2020-03-27 10:29:44,671 [pool-1-thread-1] INFO  org.perf4j.TimingLogger - start[1585304984651] time[19] tag[writeTile]
2020-03-27 10:29:44,671 [main] INFO  org.perf4j.TimingLogger - start[1585304283859] time[700812] tag[convertToPyramid]
Exception in thread "main" picocli.CommandLine$ExecutionException: Error while calling command (com.glencoesoftware.pyramid.Pyramid
FromDirectoryWriter@2892ccae): java.lang.NullPointerException
        at picocli.CommandLine.execute(CommandLine.java:1180)
        at picocli.CommandLine.access$800(CommandLine.java:141)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:1367)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:1335)
        at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:1243)
        at picocli.CommandLine.parseWithHandlers(CommandLine.java:1526)
        at picocli.CommandLine.call(CommandLine.java:1786)
        at picocli.CommandLine.call(CommandLine.java:1710)
        at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.main(PyramidFromDirectoryWriter.java:260)
Caused by: java.lang.NullPointerException
        at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.getInputTileBytes(PyramidFromDirectoryWriter.java:361)
        at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.convertToPyramid(PyramidFromDirectoryWriter.java:654)
        at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:284)
        at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:97)
        at picocli.CommandLine.execute(CommandLine.java:1173)
        ... 8 more
```

--- a/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
+++ b/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
@@ -415,14 +415,14 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
           int sizeX = metadata.getPixelsSizeX(0).getNumberValue().intValue();
           int sizeY = metadata.getPixelsSizeY(0).getNumberValue().intValue();
           if (descriptor.sizeX != sizeX) {
-              throw new FormatException(String.format(
-                  "Resolution %d dimension mismatch! metadata=%d pyramid=%d",
-                  resolution, descriptor.sizeX, sizeX));
+            throw new FormatException(String.format(
+                "Resolution %d dimension mismatch! metadata=%d pyramid=%d",
+                resolution, descriptor.sizeX, sizeX));
           }
           if (descriptor.sizeY != sizeY) {
-              throw new FormatException(String.format(
-                  "Resolution %d dimension mismatch! metadata=%d pyramid=%d",
-                  resolution, descriptor.sizeY, sizeY));
+            throw new FormatException(String.format(
+                "Resolution %d dimension mismatch! metadata=%d pyramid=%d",
+                resolution, descriptor.sizeY, sizeY));
           }
         }
       }

--- a/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
+++ b/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
@@ -113,9 +113,6 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
 
   private static final long FIRST_IFD_OFFSET = 8;
 
-  /** Scaling factor between two adjacent resolutions. */
-  private static final int PYRAMID_SCALE = 2;
-
   /** Name of label image file. */
   private static final String LABEL_FILE = "LABELIMAGE.jpg";
 
@@ -401,31 +398,33 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
       descriptor.resolutionNumber = resolution;
 
       DatasetAttributes attrs = n5Reader.getDatasetAttributes("/" + resolution);
+      descriptor.sizeX = (int) attrs.getDimensions()[0];
+      descriptor.sizeY = (int) attrs.getDimensions()[1];
       descriptor.tileSizeX = attrs.getBlockSize()[0];
       descriptor.tileSizeY = attrs.getBlockSize()[1];
       rgbChannels = attrs.getBlockSize()[2];
       descriptor.numberOfTilesX =
-        getTileCount(attrs.getDimensions()[0], descriptor.tileSizeX);
+        getTileCount(descriptor.sizeX, descriptor.tileSizeX);
       descriptor.numberOfTilesY =
-        getTileCount(attrs.getDimensions()[1], descriptor.tileSizeY);
+        getTileCount(descriptor.sizeY, descriptor.tileSizeY);
 
       if (resolution == 0) {
+        // If we have image metadata available sanity the dimensions against
+        // those in the underlying N5 pyramid
         if (metadata.getImageCount() > 0) {
-          descriptor.sizeX =
-            metadata.getPixelsSizeX(0).getNumberValue().intValue();
-          descriptor.sizeY =
-            metadata.getPixelsSizeY(0).getNumberValue().intValue();
+          int sizeX = metadata.getPixelsSizeX(0).getNumberValue().intValue();
+          int sizeY = metadata.getPixelsSizeY(0).getNumberValue().intValue();
+          if (descriptor.sizeX != sizeX) {
+              throw new FormatException(String.format(
+                  "Resolution %d dimension mismatch! metadata=%d pyramid=%d",
+                  resolution, descriptor.sizeX, sizeX));
+          }
+          if (descriptor.sizeY != sizeY) {
+              throw new FormatException(String.format(
+                  "Resolution %d dimension mismatch! metadata=%d pyramid=%d",
+                  resolution, descriptor.sizeY, sizeY));
+          }
         }
-        else {
-          descriptor.sizeX = descriptor.tileSizeX * descriptor.numberOfTilesX;
-          descriptor.sizeY = descriptor.tileSizeY * descriptor.numberOfTilesY;
-        }
-      }
-      else {
-        descriptor.sizeX =
-          resolutions.get(resolution - 1).sizeX / PYRAMID_SCALE;
-        descriptor.sizeY =
-          resolutions.get(resolution - 1).sizeY / PYRAMID_SCALE;
       }
 
       resolutions.add(descriptor);

--- a/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
+++ b/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
@@ -441,15 +441,6 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
   }
 
   /**
-   * @param resolution the resolution index
-   * @return total scale factor between the given resolution and the full
-   *         resolution image (resolution 0)
-   */
-  private double getScale(int resolution) {
-    return Math.pow(PYRAMID_SCALE, resolution);
-  }
-
-  /**
    * Populate number of channels, pixels types, endianess, etc.
    * based on the first tile
    * @throws IOException

--- a/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
+++ b/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
@@ -358,6 +358,11 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
     DataBlock block = n5Reader.readBlock(
       blockPath, n5Reader.getDatasetAttributes(blockPath),
       gridPosition);
+
+    if (block == null) {
+      throw new FormatException("Could not find block " + blockPath);
+    }
+
     ByteBuffer buffer = block.toByteBuffer();
     byte[] tile = new byte[xy * bpp * rgbChannels];
     boolean isPadded = buffer.limit() > tile.length;
@@ -400,11 +405,9 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
       descriptor.tileSizeY = attrs.getBlockSize()[1];
       rgbChannels = attrs.getBlockSize()[2];
       descriptor.numberOfTilesX =
-        (int) Math.ceil(
-            (double) attrs.getDimensions()[0] / descriptor.tileSizeX);
+        getTileCount(attrs.getDimensions()[0], descriptor.tileSizeX);
       descriptor.numberOfTilesY =
-        (int) Math.ceil(
-            (double) attrs.getDimensions()[1] / descriptor.tileSizeY);
+        getTileCount(attrs.getDimensions()[1], descriptor.tileSizeY);
 
       if (resolution == 0) {
         if (metadata.getImageCount() > 0) {
@@ -556,9 +559,13 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
       Object realY = originalMeta.get(levelKey + " | Y");
       if (realX != null) {
         descriptor.sizeX = DataTools.parseDouble(realX.toString()).intValue();
+        descriptor.numberOfTilesX =
+          getTileCount(descriptor.sizeX, descriptor.tileSizeX);
       }
       if (realY != null) {
         descriptor.sizeY = DataTools.parseDouble(realY.toString()).intValue();
+        descriptor.numberOfTilesY =
+          getTileCount(descriptor.sizeY, descriptor.tileSizeY);
       }
 
       if (descriptor.resolutionNumber == 0) {
@@ -647,6 +654,10 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
                 descriptor.tileSizeX, descriptor.sizeX - region.x);
               region.height = (int) Math.min(
                 descriptor.tileSizeY, descriptor.sizeY - region.y);
+
+              if (region.width <= 0 || region.height <= 0) {
+                continue;
+              }
 
               StopWatch t0 = new Slf4JStopWatch("getInputTileBytes");
               byte[] tileBytes;
@@ -1102,6 +1113,17 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
     if (overwrite) {
       overwriteNextOffset(offsetPointer);
     }
+  }
+
+  /**
+   * Calculate the number of tiles for a dimension based upon the tile size.
+   *
+   * @param size the number of pixels in the dimension (e.g. image width)
+   * @param tileSize the number of pixels in the tile along the same dimension
+   * @return the number of tiles
+   */
+  private int getTileCount(long size, long tileSize) {
+    return (int) Math.ceil((double) size / tileSize);
   }
 
 }

--- a/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
+++ b/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
@@ -355,7 +355,7 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
 
     String blockPath = "/" + resolution;
     long[] gridPosition = new long[] {x, y, no};
-    DataBlock block = n5Reader.readBlock(
+    DataBlock<?> block = n5Reader.readBlock(
       blockPath, n5Reader.getDatasetAttributes(blockPath),
       gridPosition);
 
@@ -454,7 +454,7 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
     interleaved = false;
     rgbChannels = 1;
     String blockPath = "/0";
-    DataBlock block = n5Reader.readBlock(blockPath,
+    DataBlock<?> block = n5Reader.readBlock(blockPath,
       n5Reader.getDatasetAttributes(blockPath), new long[] {0, 0, 0});
     littleEndian = block.toByteBuffer().order() == ByteOrder.LITTLE_ENDIAN;
     if (block instanceof ByteArrayDataBlock) {

--- a/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
+++ b/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
@@ -553,21 +553,6 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
       LOG.info("Adding metadata for resolution: {}",
         descriptor.resolutionNumber);
 
-      String levelKey =
-        "Image #0 | Level sizes #" + descriptor.resolutionNumber;
-      Object realX = originalMeta.get(levelKey + " | X");
-      Object realY = originalMeta.get(levelKey + " | Y");
-      if (realX != null) {
-        descriptor.sizeX = DataTools.parseDouble(realX.toString()).intValue();
-        descriptor.numberOfTilesX =
-          getTileCount(descriptor.sizeX, descriptor.tileSizeX);
-      }
-      if (realY != null) {
-        descriptor.sizeY = DataTools.parseDouble(realY.toString()).intValue();
-        descriptor.numberOfTilesY =
-          getTileCount(descriptor.sizeY, descriptor.tileSizeY);
-      }
-
       if (descriptor.resolutionNumber == 0) {
         MetadataTools.populateMetadata(
           this.metadata, 0, null, this.littleEndian, "XYCZT",

